packages/client-proxy/internal/edge/pool/edge-pool.go

```diff
@@ -182,36 +182,36 @@ func (p *EdgePool) connectNode(ctx context.Context, node *sd.ServiceDiscoveryIte
 		return err
 	}
 
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
 	p.nodes.Insert(o.ServiceId, o)
 
-	// call initial node sync
-	return p.syncNode(ctx, o, true)
+	return nil
 }
 
-func (p *EdgePool) syncNode(ctx context.Context, node *EdgeNode, foundWithDiscovery bool) error {
-	_, childSpan := p.tracer.Start(ctx, "sync-edge-node")
+func (p *EdgePool) removeNode(ctx context.Context, node *EdgeNode) error {
+	_, childSpan := p.tracer.Start(ctx, "remove-edge-node")
 	defer childSpan.End()
 
-	// close connection with node
-	if !foundWithDiscovery {
-		p.logger.Info("Edge node connection is not active anymore, closing.", zap.String("node_id", node.ServiceId))
+	p.logger.Info("Edge node connection is not active anymore, closing.", zap.String("node_id", node.ServiceId))
 
-		err := node.Close()
-		if err != nil {
-			p.logger.Error("Error closing connection to node", zap.Error(err))
-		}
+	err := node.Close()
+	if err != nil {
+		p.logger.Error("Error closing connection to node", zap.Error(err))
+	}
 
-		// stop background sync and close everything
-		err = node.Kill()
-		if err != nil {
-			p.logger.Error("Error closing connection to node", zap.Error(err))
-		}
+	// stop background sync and close everything
+	err = node.Kill()
+	if err != nil {
+		p.logger.Error("Error closing connection to node", zap.Error(err))
+	}
 
-		p.nodes.Remove(node.ServiceId) // remove from pool
+	p.mutex.Lock()
```

Comment:
does it make sense to lock here?  You probably don't want to use node which have already closed connection?

Reply:
You want to lock, because what if at the same time you will request it from map with listing/get function?